Fix embedded tracker

FranciscoPombal opened the pull request:
Log more request processing events, improve processing of actual + claimed address.

- ~Some of the errors in the announce request processing are now warnings~

- ~Errors and warnings in the announce request processing are now logged~

- ~Client address conversion to plain IPv4 is now logged when it happens~

Follow-up to: https://github.com/qbittorrent/qBittorrent/issues/12257

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -244,6 +245,9 @@ Http::Response Tracker::processRequest(const Http::Request &request, const Http:
     catch (const HTTPError &error) {
+        LogMsg(tr("Embedded Tracker: HTTP request error: %1 %2")
+                .arg(QString::number(error.statusCode()), error.statusText())
+            , Log::CRITICAL);
```
You misuse Log::CRITICAL. It should only be used when qbt cannot continue to operate normally.

FranciscoPombal replied:
Ok, I guess I can make everything a `WARNING` or `INFO` instead.

FranciscoPombal replied:
Fixed.

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -271,52 +275,95 @@ void Tracker::processAnnounceRequest()
-    if (ok)
+    if (ok) {
         announceReq.socketAddress = QHostAddress(decimalIPv4);
+        LogMsg(tr(R"(Embedded Tracker: Announce request from non- plain-IPv6 address %1.)"
```
I don't see why you need to use raw string literal.

FranciscoPombal replied:
Using raw string literals prevents escape hell. It's more consistent to use it everywhere, if we already use it once. What are the disadvantages?

FranciscoPombal replied:
Downvotes without explanation are not helpful.

glassez replied:
>It's more consistent to use it everywhere, if we already use it once.

This is a very destructive idea, IMO. Everything must have its place and time. It is not designed as a replacement for regular string literals.

glassez replied:
>Downvotes without explanation are not helpful.

Then they would be useless...

FranciscoPombal replied:
@glassez 
> > Downvotes without explanation are not helpful.
> 
> Then they would be useless...

Only if we start acting like redditors, which I hope doesn't happen.

> This is a very destructive idea, IMO. Everything must have its place and time. It is not designed as a replacement for regular string literals.

But for these specific usages, what is the problem? By everywhere, I meant "everywhere in the embedded tracker code", changing it everywhere in the code base might indeed not be desirable, but that's out of scope of this PR anyway.

I can understand undoing this change here if you are of the opinion that "changing to raw string literals where it makes sense should be done at once for the whole code base in another PR". Is that the case?

glassez replied:
> By everywhere, I meant "everywhere in the embedded tracker code"

It doesn't change my answer.
>I can understand undoing this change here if you are of the opinion that "changing to raw string literals where it makes sense should be done at once for the whole code base in another PR". Is that the case?

No.
I just mean that use if raw string literal has no sense in case of simple string like this one. And some kind of unification is not appropriate here.

FranciscoPombal replied:
@glassez Ok, I'll change it to normal literals where it is not needed.

FranciscoPombal replied:
Ok, now raw string literals are only used when the strings contain `"`. Do you think I should revert to using escaped strings also in that case?

FranciscoPombal replied:
@glassez ping, like I said, raw string literals are now only used when justifiable (i.e. when the string itself also contains `"`). Is this acceptable, or do you really want me to revert to using escaped strings all around?

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -271,52 +275,95 @@ void Tracker::processAnnounceRequest()
     // 1. info_hash
     const auto infoHashIter = queryParams.find(ANNOUNCE_REQUEST_INFO_HASH);
-    if (infoHashIter == queryParams.end())
-        throw TrackerError("Missing \"info_hash\" parameter");
```
perhaps we should allow exception messages to be translated.

FranciscoPombal replied:
Probably a good idea, will handle that after the raw string literals question is resolved.

Chocobo1 replied:
I retract this, error message for peers should stay in English.

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -271,52 +275,95 @@ void Tracker::processAnnounceRequest()
+    }
 
     announceReq.peer.peerId = *peerIdIter;
+    announceReq.peer.peerId.truncate(20);
```
do we really have to do this?

FranciscoPombal replied:
https://github.com/qbittorrent/qBittorrent/pull/12434/files#r404058554

Chocobo1 replied:
IMO we are wasting CPU cycles on something that has no effect (except your mind of "more leniency")...

Chocobo1 replied:
On a second thought, maybe this does make sense, I mean to adhere to BEP. 

>https://github.com/qbittorrent/qBittorrent/pull/12434/files#r404058554

This is not a constructive response. I see you're even lazy to state your reason for the change.

Chocobo1 replied:
you'll need to use `peerIdIter->left(20)` on previous line.

FranciscoPombal replied:
@Chocobo1
> This is not a constructive response. I see you're even lazy to state your reason for the change.

Not lazy, I only linked the response because the response would have been the exact same as for other similar comments... would you rather I had copy-pasted the exact same thing multiple times? These are all instances of apparently not-so-compliant cases where the spec does not explicitly mandate the request to fail, and where we can default to a reasonable lenient behavior.

But just to be extremely clear in this specific case:
The spec says the length is 20, so we truncate to that internally (while still logging if this happens), which is a reasonable lenient alternative to just not processing the request at all if the length is not 20, because the spec does not explicitly say "if `peer_id` is not 20 bytes in length, the request should/must fail". In fact, it does not explicitly state what to do at all when the length is not 20.

FranciscoPombal replied:
@Chocobo1

> it is better to use `peerIdIter->left(20)` on previous line.

Fine by me. Out of curiosity, what are the benefits?

FranciscoPombal replied:
Fixed, by the way.

Chocobo1 replied:
Isn't it obvious? so it would copy a shorter array instead of a longer one.

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -149,6 +149,7 @@ using namespace BitTorrent;
 struct Tracker::TrackerAnnounceRequest
 {
     QHostAddress socketAddress;
+    quint16 requestPort = 0;
```
I don't see why you need a field in this struct.

FranciscoPombal replied:
Sure I could make it a variable because in this PR specifically it is only used in one scope. But in terms of design, and for future modifications, doesn't it make sense that the port number be part of this struct? The address is already there as well.

Chocobo1 replied:
>doesn't it make sense that the port number be part of this struct? 

No, at least not in this PR.

FranciscoPombal replied:
Ok, since you insist, and I really don't have a specific more compelling reason to do it the other way, I'll change it.

FranciscoPombal replied:
Fixed, and I also changed an instance of `ushort` to `quint16`, since that's what we use elsewhere.

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -357,7 +404,11 @@ void Tracker::processAnnounceRequest()
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::WARNING);
+        announceReq.event = ANNOUNCE_REQUEST_EVENT_EMPTY;
```
why? I don't think this make sense.

FranciscoPombal replied:
Again, why is more leniency here not desirable? A warning will still be logged.

Chocobo1 replied:
I don't think you really answered my question...

glassez replied:
>Invalid "event" parameter, assuming "empty"

Is this behavior a part of protocol standard? If no I strongly object to encourage invalid clients! All invalid requests should be dropped.

FranciscoPombal replied:
@glassez BEP 3 does not specify what should be done in the case of an invalid value for this parameter (i.e. it does not explicitly say "a request containing any other values should/must be considered invalid"). So I think it is better to be lenient, and just log a warning.Otherwise, we can just keep the logging statement explaining that the request was dropped.

FranciscoPombal replied:
@arvidn Can we get your opinion on this matter? In this case, and also the `numwant` discussion below, is it better to be lenient and attempt serve the request if possible, falling back to a sane default, or to drop the request altogether?

arvidn replied:
without having thought about this case very much, generally I think it's a good default to be strict. (imagine if all bittorrent clients had required bencoded dictionaries to be sorted correctly for instance, clients wouldn't have to go out of their way to support invalid encodings as they do today).

Maybe some relevant questions for this case. (1) a client that sends an invalid request to the tracker, is it likely to be a correct about other arguments, say the info-hash? (2) is it the kind of peer that a tracker should hand out to ther peers?

FranciscoPombal replied:
@arvidn I see.
1, 2 - Probably not, on both counts. So I guess being strict is the better option.

But in any case, shouldn't the spec be updated to be more strict? At least the V2 spec, as updating the original one is not meaningful at this point. If the spec is not strict enough, it allows implementations to make assumptions that could lead to ossification of the protocol - the most well-known case of this is TLS 1.x -> 1.3 I think. Isn't it desirable to prevent that?

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -271,52 +275,95 @@ void Tracker::processAnnounceRequest()
+                , Log::WARNING);
+        }
+        else { 
+            announceReq.numwant = num;
```
:\ This won't do you anything good.

FranciscoPombal replied:
Why not? Why is more leniency here not desirable? A warning will still be logged.

Chocobo1 replied:
What if I really want to set it to a negative value and does not want any peer in list? then you would give the unexpected result, no?

glassez replied:
If negative value isn't acceptable here then the entire request is invalid (unless any fallback is defined by protocol).

Chocobo1 replied:
>If negative value isn't acceptable here then the entire request is invalid (unless any fallback is defined by protocol).

Well, the standard is really obscure on this, IMO this gives room for interpretation. Personally I would treat it as-is and return no peer.

FranciscoPombal replied:
@Chocobo1
In that case you just set numwant = 0, same as before. The difference now is that negative values will not result in an error.

@glassez 
https://github.com/qbittorrent/qBittorrent/pull/12434#discussion_r404116723

Chocobo1 replied:
>In that case you just set numwant = 0, same as before. 

Is it really that hard to understand that people/hackers will try to use any value here? You give unexpected results and you just hand waving my concerns? (you do it all the time)
I'm sorry I cannot continue any more. Do as you want, this has nothing to do with me.

FranciscoPombal replied:
@Chocobo1

Calm down please, I'm not hand waving anything. The `numwant` field defaults to `50`, so if we don't reassign it, it stays at 50, which is safe.

Previous behavior:

- `num_want > 0`: success.
- `num_want == 0`: success (no peer list returned).
- `num_want < 0`: failure for the announcer and silent failure for the tracker "admin".


New behavior:

- `num_want > 0`: same as before.
- `num_want == 0`: same as before.
- `num_want < 0`: success for the announcer, using our default value for `numwant` of `50`. The tracker "admin" is warned that an announcer has supplied an invalid number, but a default value was used instead to still be able to service the request.

Would you rather I change the text "falling back to default" to "using default value instead" or "using default value of 50 instead" or "using default value of 50"?

Chocobo1 replied:
>If negative value isn't acceptable here then the entire request is invalid (unless any fallback is defined by protocol).

Just to clarify my stance on this, @glassez is right, invalid parameter should stop processing (to avoid wasting more resources) and send back the error message to the peer.

FranciscoPombal replied:
The thing is, while the protocol does not define a specific fallback, it does not say the requests should immediately be dropped either.

arvidn replied:
it's not obvious to me that `&numwant=-1` is any more correct (or should be accepted to any greater degree) than `&numwant=foo`.

As far as I can tell (but I haven't thought too much about this either), the only useful values for `numwant` is `0` or a *positive* number.  Clients will generally want as many peers as possible, or 0 (if they are stopping for instance). And the actual number to send has to be up to the tracker to decide.

FranciscoPombal replied:
Alright, I removed the leniency changes.

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -271,52 +275,95 @@ void Tracker::processAnnounceRequest()
-    if (ok)
+    if (ok) {
         announceReq.socketAddress = QHostAddress(decimalIPv4);
+        LogMsg(tr(R"(Embedded Tracker: Announce request from non- plain-IPv6 address %1.)"
```
You call LogMsg here and at exception handler, what a mess...
At this point this PR is beyond my comprehension, sorry.

FranciscoPombal replied:
This is not constructive criticism. Do you mean I should remove this logging statement specifically, or should I remove the `LogMsg` call from the `HTTPError`, or both? If so, why?

Chocobo1 replied:
Sorry, all of this is over my head, I couldn't even give a proper response...

FranciscoPombal replied:
@Chocobo1

I mean, am I doing something really bad or what? I just wanted a bit more logging overall, am I doing it in the wrong places in general?

In this specific place and the `HTTPError` exception handling, the behavior doesn't change at all, the only difference is the logging statements.

glassez commented:
https://github.com/qbittorrent/qBittorrent/pull/12043#pullrequestreview-363194074

FranciscoPombal commented:
@glassez 

> [#12043 (review)](https://github.com/qbittorrent/qBittorrent/pull/12043#pullrequestreview-363194074)

Fixed.

FranciscoPombal commented:
@Chocobo1 I removed the leniency changes. Now this is just about logging more stuff in the embedded tracker.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -271,51 +274,93 @@ void Tracker::processAnnounceRequest()
-    if (ok)
+    if (ok) {
         announceReq.socketAddress = QHostAddress(decimalIPv4);
+        LogMsg(tr("Embedded Tracker: Announce request from non- plain-IPv6 address %1."
```
Due to comment above this message is incorrect to be here since this place can be reached in case of plain IPv4 address.

FranciscoPombal replied:
@glassez 
This message is supposed to be printed every time anything _except_ a **plain IPv6** is used. This includes the case of a **plain IPv4**.
In other words, this message prints "if address is indeed IPv4 or if it is an IPv4-mapped IPv6 address", which is what the comment above says.

The point of this is to help track and diagnose connectivity issues with the embedded tracker when IPv4-mapped IPv6 addresses are used and possibly other cases.

glassez replied:
>This includes the case of a plain IPv4.

But it is a trivial case. Why should it be logged (unlike plain IPv6)?

FranciscoPombal replied:
@glassez because I did not find an easy way to exclude it.

glassez replied:
Can't you check whether it was IPv4 in the first place (before it was adjusted)?

FranciscoPombal replied:
@glassez Of course I can. Brain freeze, sorry. It's fixed now.

FranciscoPombal commented:
@glassez 
> I wouldn't merge such changes until Logger is redesigned to level-based form. Otherwise, it can generate a whole bunch of messages that can't be filtered out (I mean completely ignored, not only removed from the log list in GUI, as we allow now).
> IMO, such messages should be of Debug level since they indicates client errors that won't affect the app workability.

We don't know how long it will take until the logging is redesigned. It could take a long time. In the meantime, I think this is a relatively small change that can come in very handy when diagnosing issues with the embedded tracker. Otherwise, every time I try to solve a problem with the embedded tracker, I basically have to ask the bug reporter for a wireshark capture (not a realistic request for the majority of the users). With this change, hopefully solve the issue with just the information in the logs.

Plus in terms of code structure this is pretty trivial to change/undo when the logging is finally redesigned.

glassez commented:
>We don't know how long it will take until the logging is redesigned. It could take a long time. 

It can be done quickly as soon as we reach any agreement on it. But for some reason this topic is ignored...

glassez commented:
>Plus in terms of code structure this is pretty trivial to change/undo when the logging is finally redesigned.

Well, I'll approve it unless my objections are supported by someone else.

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -244,6 +244,9 @@ Http::Response Tracker::processRequest(const Http::Request &request, const Http:
     catch (const HTTPError &error) {
+        LogMsg(tr("Embedded Tracker: HTTP request error: %1 %2")
+                .arg(QString::number(error.statusCode()), error.statusText())
+            , Log::WARNING);
```
Log::WARNING is misused. Any error from client isn't relevant to qbt normal operations.

FranciscoPombal replied:
fixed

Chocobo1 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -271,51 +274,95 @@ void Tracker::processAnnounceRequest()
+    const quint16 requestPort =  m_env.clientPort;
+
+    // Enforce using IPv4 if address is an IPv4-mapped IPv6 address
+    if(announceReq.socketAddress.protocol() == QAbstractSocket::IPv6Protocol) {
```
coding style.

FranciscoPombal replied:
fixed

glassez commented:
>And before you ask, currently I don't have opinions about the redesign of logger.

Do you have any idea what I'm talking about (have you read my suggestion about it elsewhere)? If so, does your answer here mean that you are simply neutral about it, so that you have no fundamental objection to changing it?

Chocobo1 commented:
>Do you have any idea what I'm talking about (have you read my suggestion about it elsewhere)? 

No, I don't have a complete picture, please remind me about it.

FranciscoPombal commented:
@Chocobo1
> IMO logging client error is out of the scope for a (small) embedded server. The current server already sent back proper error message to offending clients and I think the debugging (and therefore logging) should happen at the client side (as that is where the error originates, at least for the errors you're logging in this PR).
> Personally I don't think this PR is a good idea and won't support it, but I don't mind if other devs agree with this PR (i.e. there will be no hard block from me).
> 
> And before you ask, currently I don't have opinions about the redesign of logger.

I thought about it some more, and I think you are mostly right. I will remove the logging of the request params errors, as it is somewhat redundant at the end of the day. However, I still think logging the IP/protocol stuff at the tracker side is very useful for diagnosing potential problems. The potential IPv4-mappend IPv6 -> IPv4 conversion is information that is not available anywhere else short of debugging the program, for example.

glassez commented:
>No, I don't have a complete picture, please remind me about it.

I suggest to use log message types based on verbosity levels instead of current "flat" (and mostly confusing) typification:
1. We define several verbosity levels, e.g. (starting from most verbose) Debug, Info, Warning, Critical.
2. We assign level to log message according to its relevance.
3. We allow to set Log level threshold via settings.
4. Log just ignores all messages that have more verbose level that configured threshold (i.e. it doesn't store such messages and it doesn't notify about them). E.g. if "Warning" level is set as threshold only "Warning" and "Critical" messages will be processed.

With this architecture, we can easily add more messages (even diagnostic ones) without fear of overloading the productive configuration (which is supposed to have a threshold higher than Debug).

Chocobo1 commented:
In general, I feel like it is quite similar to what we are using currently.

>E.g. if "Warning" level is set as threshold only "Warning" and "Critical" messages will be processed.

As a user, I would like to have the ability to select each log level independently.

As for your earlier comment:
>Otherwise, it can generate a whole bunch of messages that can't be filtered out (I mean completely ignored, not only removed from the log list in GUI, as we allow now).

Currently logger stores all log messages without filtering them, I don't mind we change this.

glassez commented:
>As a user, I would like to have the ability to select verbosity levels independently.

Seems you don't understand it. From the Log perspective any level contains messages from all previous (less verbose) levels. It seems meaningless to store, e.g., only "Debug" messages.
But we can preserve current filtering at the UI side so user can see only selected message types (of course from the subset that is provided by logger according to selected threshold).

Chocobo1 commented:
>Seems you don't understand it. From the Log perspective any level contains messages from all previous (less verbose) levels. It seems meaningless to store, e.g., only "Debug" messages.

I simply don't like to be forced to show levels that I don't want. What if I really want to suppress warning level but still want info & critical levels? The verbosity hierarchy (or threshold) you presented is quite limited and inflexible from a user viewpoint.

>With this architecture, we can easily add more messages (even diagnostic ones) without fear of overloading the productive configuration 

I admit I don't understand what this has to do with verbosity hierarchy. Basically you just need to filter the messages according to its level at `Logger::addMessage()` and add an off by default `debug` level.

And that is enough about my opinion, maybe you should wait for others opinion too (I really don't mind changing it if the majority agrees).

Chocobo1 commented:
>But we can preserve current filtering at the UI side so user can see only selected message types (of course from the subset that is provided by logger according to selected threshold).

If this is possible then I suppose I'm fine with it.

glassez commented:
>I simply don't like to be forced to show levels that I don't want. What if I really want to suppress warning level but still want info & critical levels?

I talk about core Log architecture. And I didn't invent it, it's widely used (to be honest, I've always wondered where qBittorrent log structuring came from). It even seems to follow the general principle of verbosity, used not only in logging. From this perspective, it is useless to cut out any particular layer of messages.
>What if I really want to suppress warning level but still want info & critical levels?

What's the point to store, e.g., only Info messages? What can you use them for?
The point is to tell the Logger to be more verbose (or less).
E.g. a productive configuration usually does not require all messages, especially Debug ones. The debug configuration can include all of them. For example, when something starts to go wrong, the user can sacrifice performance to be able to provide more debugging information.

>The verbosity hierarchy (or threshold) you presented is quite limited and inflexible from a user viewpoint.

Looks like you're talking about a View layer after all. So my proposal is still here:
>we can preserve current filtering at the UI side so user can see only selected message types (of course from the subset that is provided by logger according to selected threshold).

arvidn commented:
I think the main use case for logging is to debug and troubleshoot issues. I think when enabling more verbose logging, it's almost always to trouble shoot a specific issue. Being able to narrow down which component(s) print additional logging can save a lot of log noise from unrelated parts of the program. Unrelated to the issue being trouble shot that is.

This has worked well in uTorrent (where the log tab context menu lets you check categories of logs to enable).

FranciscoPombal commented:
@Chocobo1 @glassez 
Alright, I've addressed the comments and made a few changes to improve the embedded tracker implementation in general.

- Both the actual address and the claimed address are treated similarly now, including converting to IPv4 if they are IPv4-mapped IPv6
- From the beginning, it is guaranteed that the addresses are either pure IPv4 or pure IPv6, in order to prevent surprises with converting to QByteArray, for example
- Other minor things
- Includes only the logging changes previously agreed upon.

FranciscoPombal commented:
@glassez @Chocobo1 

Rebased on top of `master`. Thoughts on https://github.com/qbittorrent/qBittorrent/pull/12434#issuecomment-617327777?
